sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -868,6 +866,42 @@ object TransposeWindow extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Infers filters from [[Generate]], such that rows that would have been removed
+ * by this [[Generate]] can be removed earlier - before joins and in data sources.
+ */
+object InferFiltersFromGenerate extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    // Don't infer filters from foldable expressions to avoid
+    // constant filters like 'size([1, 2, 3]) > 0'. These can add
+    // unnecessary overhead and these will not show up in the childs
+    // constraints, breaking this rules idempotency.
```

Comment:
nit: How about rephrasing it like this?
```
    // This rule does not infer filters from foldable expressions to avoid constant filters
    // like 'size([1, 2, 3]) > 0'. These do not show up in child's constraints and
    // then the idempotence will break.
```